rkt api service panics

```
2015/12/17 10:15:10 API service starting...
2015/12/17 10:15:10 API service running on localhost:15441...
2015/12/17 10:21:15 Failed to read pod manifest for pod "1ffeb0be-88f9-4559-b4d9-06b918313e1c": no such file or directory
2015/12/17 10:21:16 Failed to read pod manifest for pod "1ffeb0be-88f9-4559-b4d9-06b918313e1c": no such file or directory
2015/12/17 10:21:17 Failed to read pod manifest for pod "1ffeb0be-88f9-4559-b4d9-06b918313e1c": no such file or directory
2015/12/17 10:21:17 Failed to read pod manifest for pod "1ffeb0be-88f9-4559-b4d9-06b918313e1c": no such file or directory
panic: cas db lock already gained
```

Full trace here:
https://gist.github.com/yifan-gu/a0cca9884bbda373a20c

--- a/store/db.go
+++ b/store/db.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"sync"
 
 	"github.com/coreos/rkt/pkg/lock"
 
@@ -29,9 +30,42 @@ const (
 	DbFilename = "ql.db"
 )
 
+// dbLock is used to guarantee both thread-safety and process-safety
+// for db access.
+type dbLock struct {
+	fl *lock.FileLock
+	// Although the FileLock already ensures thread safety, the Go runtime is unaware
+	// of this, and so Mutex is necessary to satisfy the Go race detector.
+	sync.Mutex
+}
+
+func newDBLock(dirPath string) (*dbLock, error) {
+	l, err := lock.NewLock(dirPath, lock.Dir)
+	if err != nil {
+		return nil, err
+	}
+	return &dbLock{fl: l}, nil
+}
+
+func (dl *dbLock) lock() error {
+	if err := dl.fl.ExclusiveLock(); err != nil {
+		return err
+	}
+	dl.Lock()
+	return nil
+}
+
+func (dl *dbLock) unlock() error {
+	if err := dl.fl.Unlock(); err != nil {
+		return err
+	}
+	dl.Unlock()
+	return nil
+}
+
 type DB struct {
 	dbdir string
-	lock  *lock.FileLock
+	dl    *dbLock
 	sqldb *sql.DB
 }
 
@@ -39,44 +73,45 @@ func NewDB(dbdir string) (*DB, error) {
 	if err := os.MkdirAll(dbdir, defaultPathPerm); err != nil {
 		return nil, err
 	}
-	return &DB{dbdir: dbdir}, nil
+
+	dl, err := newDBLock(dbdir)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DB{dbdir: dbdir, dl: dl}, nil
 }
 
 func (db *DB) Open() error {
-	// take a lock on db dir
-	if db.lock != nil {
-		panic("cas db lock already gained")
-	}
-	dl, err := lock.ExclusiveLock(db.dbdir, lock.Dir)
-	if err != nil {
+	if err := db.dl.lock(); err != nil {
 		return err
 	}
-	db.lock = dl
 
 	sqldb, err := sql.Open("ql", filepath.Join(db.dbdir, DbFilename))
 	if err != nil {
-		dl.Close()
+		unlockErr := db.dl.unlock()
+		if unlockErr != nil {
+			err = fmt.Errorf("[%s, %s]", err, unlockErr)
+		}
 		return err
 	}
 	db.sqldb = sqldb
+
 	return nil
 }
 
 func (db *DB) Close() error {
-	if db.lock == nil {
-		panic("cas db, Close called without lock")
-	}
 	if db.sqldb == nil {
 		panic("cas db, Close called without an open sqldb")
 	}
+
 	if err := db.sqldb.Close(); err != nil {
 		return fmt.Errorf("cas db close failed: %v", err)
 	}
 	db.sqldb = nil
 
-	db.lock.Close()
-	db.lock = nil
-	return nil
+	// Don't close the flock as it will be reused.
+	return db.dl.unlock()
 }
 
 func (db *DB) Begin() (*sql.Tx, error) {

--- a/store/db_test.go
+++ b/store/db_test.go
@@ -1,0 +1,117 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"testing"
+
+	"github.com/coreos/rkt/tests/testutils"
+)
+
+func queryValue(query string, tx *sql.Tx) (int, error) {
+	var value int
+	rows, err := tx.Query(query)
+	if err != nil {
+		return -1, err
+	}
+	defer rows.Close()
+
+	if !rows.Next() {
+		return -1, fmt.Errorf("no result of %q", query)
+	}
+	if err := rows.Scan(&value); err != nil {
+		return -1, err
+	}
+	return value, nil
+}
+
+func insertValue(db *DB) error {
+	return db.Do(func(tx *sql.Tx) error {
+		// Get the current count.
+		count, err := queryValue("SELECT count(*) FROM rkt_db_test", tx)
+		if err != nil {
+			return err
+		}
+		// Increase the count.
+		_, err = tx.Exec(fmt.Sprintf("INSERT INTO rkt_db_test VALUES (%d)", count+1))
+		return err
+	})
+}
+
+func getMaxCount(db *DB, t *testing.T) int {
+	var maxCount int
+	var err error
+	if err := db.Do(func(tx *sql.Tx) error {
+		// Get the maximum count.
+		maxCount, err = queryValue("SELECT max(counts) FROM rkt_db_test", tx)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to get the maximum count: %v", err)
+	}
+	return maxCount
+}
+
+func createTable(db *DB, t *testing.T) {
+	if err := db.Do(func(tx *sql.Tx) error {
+		_, err := tx.Exec("CREATE TABLE rkt_db_test (counts int)")
+		return err
+	}); err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+}
+
+func TestDBRace(t *testing.T) {
+	oldGoMaxProcs := runtime.GOMAXPROCS(runtime.NumCPU())
+	defer runtime.GOMAXPROCS(oldGoMaxProcs)
+
+	dir, err := ioutil.TempDir("", "rkt_db_test")
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer os.RemoveAll(dir)
+
+	db, err := NewDB(dir)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	// Create the table.
+	createTable(db, t)
+
+	// Insert values concurrently.
+	ga := testutils.NewGoroutineAssistant(t)
+	runs := 100
+	ga.Add(runs)
+	for i := 0; i < runs; i++ {
+		go func() {
+			if err := insertValue(db); err != nil {
+				ga.Fatalf("Failed to insert value: %v", err)
+			}
+			ga.Done()
+		}()
+	}
+	ga.Wait()
+
+	// Check the final values.
+	maxCount := getMaxCount(db, t)
+	if maxCount != runs {
+		t.Errorf("Expected: %v, saw: %v", runs, maxCount)
+	}
+}

